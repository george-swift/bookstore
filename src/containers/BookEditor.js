/* eslint-disable jsx-a11y/label-has-associated-control */
import { useState } from 'react';
import { IoCloudUploadSharp } from 'react-icons/io5';
import { useDispatch, useSelector } from 'react-redux';
import { CATEGORIES } from '../constants';
import { editBookRequested } from '../actions';
import { getPendingUpdate, updateCompleted } from '../selectors';
import Spinner from '../components/Spinner';

export default function BookEditor() {
  const { isLoading } = useSelector((state) => state.notifications);
  const status = useSelector((state) => updateCompleted(state));
  const dispatch = useDispatch();
  const book = useSelector((state) => getPendingUpdate(state));

  const [state, setState] = useState({
    title: book.title,
    category: book.category,
    author: book.author,
    chapter: book.chapter,
    percentage: book.percentage,
  });

  const handleChange = (e) => {
    const { name, value } = e.target;
    setState({
      ...state,
      [name]: value,
    });
  };

  const resetForm = () => setState({ title: '', category: '' });

  const handleSubmit = (e) => {
    e.preventDefault();
    const { id } = book;
    dispatch(editBookRequested({ id, ...state }));
    resetForm();
  };

  return (
    <div className="editor">
      {
        status ? (
          <a href="/" className="btn btn-outline-secondary w-auto">Back to Homepage</a>
        ) : (
          <form className="row g-2 shadow" onSubmit={handleSubmit}>
            <h3 className="text-center">
              Update in Bookstore
              <Spinner loading={isLoading} />
            </h3>
            <div className="col-md-12 mb-3">
              <label htmlFor="author">Book Title *</label>
              <input
                type="text"
                className="form-control"
                name="title"
                minLength="5"
                value={state.title}
                onChange={handleChange}
                required
              />
            </div>
            <div className="col-md-12 mb-3">
              <label htmlFor="author">Category *</label>
              <select
                className="form-select"
                name="category"
                value={state.category}
                onChange={handleChange}
                required
              >
                {
                CATEGORIES.map((category) => (
                  <option key={category} value={category}>
                    {category}
                  </option>
                ))
              }
              </select>
            </div>
            <div className="col-md-12 mb-3">
              <label htmlFor="author">Author</label>
              <input
                type="text"
                className="form-control"
                name="author"
                minLength="2"
                value={state.author || ''}
                onChange={handleChange}
                required
              />
            </div>
            <div className="col-sm-6 mb-3">
              <label htmlFor="chapter">Current Chapter</label>
              <input
                type="number"
                min="0"
                className="form-control"
                name="chapter"
                value={state.chapter || ''}
                onChange={handleChange}
                required
              />
            </div>
            <div className="col-sm-6 mb-3">
              <label htmlFor="percentage">(%) Read</label>
              <input
                type="number"
                min="0"
                max="100"
                className="form-control"
                name="percentage"
                value={state.percentage || ''}
                onChange={handleChange}
                required
              />
            </div>
            <div className="col-md-12">
              <button type="submit" className="btn btn-primary w-100 mb-3">
                <span className="me-2">Save Details</span>
                <IoCloudUploadSharp />
              </button>
              <a href="/" className="btn btn-outline-secondary w-100">Back to Homepage</a>
            </div>
          </form>
        )
      }
    </div>
  );
}
