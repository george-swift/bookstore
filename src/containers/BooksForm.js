import { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { CATEGORIES } from '../constants';
import { createBookRequested } from '../actions';

const BooksForm = ({ createBookRequested }) => {
  const [state, setState] = useState({
    title: '',
    category: '',
  });

  const { title, category } = state;

  const handleChange = (e) => {
    const defaultValue = 'Category';
    const { name, value } = e.target;
    if (value === defaultValue) return;
    setState({
      ...state,
      [name]: value,
    });
  };

  const resetForm = () => setState({ title: '', category: '' });

  const handleSubmit = (e) => {
    e.preventDefault();
    if (category === '') return;
    createBookRequested({
      title,
      category,
    });
    resetForm();
  };

  return (
    <div className="container-fluid wrap-form">
      <h3 className="form-title">Add new book</h3>
      <form className="row bookform" onSubmit={handleSubmit}>
        <div className="col-lg-6">
          <input
            type="text"
            className="custom-width form-control"
            name="title"
            placeholder="Book title"
            onChange={handleChange}
            value={title}
            minLength="5"
            required
          />
        </div>
        <div className="col-lg-4 category-input">
          <select
            className="form-select"
            name="category"
            value={category}
            onChange={handleChange}
            required
          >
            <option>Category</option>
            {
              CATEGORIES.map((category) => (
                <option key={category} value={category}>{category}</option>
              ))
            }
          </select>
        </div>
        <div className="col-lg-2 submit-form">
          <button type="submit" className="btn btn-primary">ADD BOOK</button>
        </div>
      </form>
    </div>
  );
};

BooksForm.propTypes = {
  createBookRequested: PropTypes.func.isRequired,
};

export default connect(null, { createBookRequested })(BooksForm);
export { BooksForm as UnconnectedBooksForm };
