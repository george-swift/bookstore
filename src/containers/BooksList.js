import { useSelector, useDispatch } from 'react-redux';
import { removeBook, filterByCategory } from '../actions';
import { booksByCategory } from '../selectors';
import CategoryFilter from '../components/CategoryFilter';
import Book from '../components/Book';

export default function BooksList() {
  const books = useSelector((state) => booksByCategory(state));
  const dispatch = useDispatch();
  const handleRemoveBook = (book) => dispatch(removeBook(book));
  const handleFilterChange = (filter) => dispatch(filterByCategory(filter));

  return (
    <div className="container-fluid bookslist">
      <CategoryFilter filterCategory={handleFilterChange} />
      <div className="row">
        {
          books?.map((book) => (
            <Book key={book.id} book={book} removeBook={handleRemoveBook} />))
            ?? (<div className="col-12"><p>No available books in this category</p></div>)
          }
      </div>
      <hr id="section-divider" />
    </div>
  );
}
