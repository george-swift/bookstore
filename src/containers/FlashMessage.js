import { useSelector } from 'react-redux';
import { IoWarningSharp } from 'react-icons/io5';
import { updateCompleted } from '../selectors';

const FlashMessage = () => {
  const { error } = useSelector((state) => state.notifications);
  const status = useSelector((state) => updateCompleted(state));

  return (
    <>
      {
        error !== null && (
          <div className="alert alert-danger" role="alert">
            <IoWarningSharp className="mb-1" />
            <span className="ms-2">{ error }</span>
          </div>
        )
      }
      {
        status && (
          <div className="alert alert-success" role="alert">
            Book successfully uploaded!
          </div>
        )
      }
    </>
  );
};

export default FlashMessage;
