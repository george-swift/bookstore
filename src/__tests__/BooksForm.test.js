import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { render, within } from '@testing-library/react';
import { UnconnectedBooksForm } from '../containers/BooksForm';

describe('Testing the BooksForm component', () => {
  const stubAction = () => ({
    type: 'CREATE_BOOK_REQUESTED',
    payload: {},
  });

  beforeEach(() => render(<UnconnectedBooksForm createBookRequested={stubAction} />));

  it('should render correctly when passed props', () => {
    const tree = renderer
      .create(<UnconnectedBooksForm createBookRequested={stubAction} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render the add book form', () => {
    const formWrapper = document.querySelector('.wrap-form');
    const formH3 = within(formWrapper).getByText('Add new book');
    expect(formH3).toBeInTheDocument();
  });

  it('should have an input field for book title in the form', () => {
    const form = document.querySelector('.bookform');
    const titleInput = within(form).getByPlaceholderText('Book title');
    expect(titleInput).toBeInTheDocument();
  });

  it('should have a select element with book categories', () => {
    const formSelect = document.querySelector('.form-select');
    const defaultValue = within(formSelect).getByText('Category');
    expect(defaultValue).toBeInTheDocument();
    const categoryOption = within(formSelect).getByText('Sci-Fi');
    expect(categoryOption).toBeInTheDocument();
  });
});
