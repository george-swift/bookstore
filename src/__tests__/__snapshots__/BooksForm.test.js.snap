// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the BooksForm component should render correctly when passed props 1`] = `
<div
  className="container-fluid wrap-form"
>
  <h3
    className="form-title"
  >
    Add new book
  </h3>
  <form
    className="row bookform"
    onSubmit={[Function]}
  >
    <div
      className="col-lg-6"
    >
      <input
        className="custom-width form-control"
        minLength="5"
        name="title"
        onChange={[Function]}
        placeholder="Book title"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="col-lg-4 category-input"
    >
      <select
        className="form-select"
        name="category"
        onChange={[Function]}
        required={true}
        value=""
      >
        <option>
          Category
        </option>
        <option
          value="Action"
        >
          Action
        </option>
        <option
          value="Biography"
        >
          Biography
        </option>
        <option
          value="History"
        >
          History
        </option>
        <option
          value="Horror"
        >
          Horror
        </option>
        <option
          value="Kids"
        >
          Kids
        </option>
        <option
          value="Learning"
        >
          Learning
        </option>
        <option
          value="Sci-Fi"
        >
          Sci-Fi
        </option>
      </select>
    </div>
    <div
      className="col-lg-2 submit-form"
    >
      <button
        className="btn btn-primary"
        type="submit"
      >
        ADD BOOK
      </button>
    </div>
  </form>
</div>
`;
